# PrepareCodonComposition
Prepare input file to codon composition
Plasmodium Falciparum 3D7 genome data in fasta format and annotation data in gff format version 55 can be downloaded from the plasmodb.org website. Coding sequences (CDS) position data in each gene is retrieved from annotation data and then merged to one gene. Then, for each gene, a bed file of CDS is generated. The fasta of CDS genes is then generated by applying a bed file to extract sequences from genome data. 
	A SQLite database of new positions is constructed by validating each position in the annotation data to ensure that only positions obtained from CDS are used (1,896 positions), and then calculating the shortest distance between each position and each gene of annotation. The entire workflow is handled by an preparecodoncomposition.py script.	
	Finally, employ a bed file, a fasta file, and a SQLite file to enter into the codon composition software (https://github.com/mnsmar/ribothrypsis/tree/master/codon_composition) in generate a plot. This software generates two files, one with the real and one with the random count of each codon. Using an ks.py script, this data is used to calculate the K-S test and evaluate the p-values of the two groups.


	
